Add optional systemd unit

terencode opened the pull request:
~~While it might not seem useful at first, it will be if it ships with an optional systemd unit file because the user will just have to change the options and enable the unit.~~

terencode commented:
Here is an example PKGBUILD taking this MR into account:
The original behaviour is preserved but now the user can set up its config and enable the systemd user unit.
```
pkgname=jack-matchmaker-git
pkgver=0.7.1.r7.g519f280
pkgrel=1
pkgdesc="A command line tool to auto-connect JACK ports matching given patterns. Git version."
arch=('any')
url="https://github.com/SpotlightKid/jack-matchmaker"
license=('GPL2')
depends=('python-setuptools' 'jack')
conflicts=('jack-matchmaker')
source=("jack-matchmaker-git::git+https://github.com/terencode/jack-matchmaker.git")
sha256sums=('SKIP')
pkgver() {
  cd "${srcdir}/${pkgname}"
  git describe --long --tags | sed 's/\([^-]*-g\)/r\1/;s/-/./g'
}


build() {
  cd "${srcdir}/${pkgname}"

  python setup.py build
}

package() {
  cd "${srcdir}/${pkgname}"

  python setup.py install --skip-build --root="${pkgdir}" --prefix=/usr \
    --optimize=1

  install -Dm 644 example.conf $pkgdir/etc/jack-matchmaker/jack-matchmaker.conf.default
  install -Dm 644 jack-matchmaker.service -t $pkgdir/usr/lib/systemd/user/
}
```

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -169,6 +186,10 @@ def reg_callback(self, port_id, action, *args):
+    def error_callback(self, error):
+        error = error.decode('utf-8')
+        log.info(error)
+
```
This change, along the with changes to `jacklib.py`, should go into a separate PR.

SpotlightKid commented on jackmatchmaker/jacklib.py:
```diff
@@ -1262,6 +1263,15 @@ def get_time():
 jlib.jack_free.argtypes = [c_void_p]
 jlib.jack_free.restype = None
 
+jlib.jack_set_error_function.argtypes = [JackErrorCallback]
```
* `jack_set_error_function` is part of the optional API, so you need to wrap this in a try/except:

       try:
           jlib.jack_set_error_function.argtypes = ...
           jlib.jack_set_error_function.restype = None
       except AttributeError:
           jlib.jack_set_error_function = None

* `jack_set_error_function` returns `void`, so you need to add

       jlib.jack_set_error_function.restype = None

SpotlightKid commented on jackmatchmaker/jacklib.py:
```diff
@@ -1262,6 +1263,15 @@ def get_time():
+        jlib.jack_set_error_function(_error_callback)
+
+    return -1
 
```
Please run 'flake8' on all changed Python files to check for coding-style errors, such as the missing empty line here.

SpotlightKid commented on jackmatchmaker/jacklib.py:
```diff
@@ -1262,6 +1263,15 @@ def get_time():
+
+def set_error_callback(error_callback) :
+    if jlib.jack_set_error_function:
+        global _error_callback
```
The `_error_callback` should be initialized to `None` at the module global namespace.

Also, please put `global _error_callback` at the start of the function body.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
     ap.add_argument('-m', '--max-attempts', type=posnum, default=0, metavar="NUM",
                     help="Max. number of attempts to connect to JACK server (default: 0=infinite)")
-    ap.add_argument('-v', '--verbose', action="store_true", help="Be verbose")
-    ap.add_argument('--version', action='version', version='%%(prog)s %s' % __version__)
```
Why did you remove the `--version` argument?

terencode replied:
Rebasing mistake...

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
                     help="Max. number of attempts to connect to JACK server (default: 0=infinite)")
-    ap.add_argument('-v', '--verbose', action="store_true", help="Be verbose")
-    ap.add_argument('--version', action='version', version='%%(prog)s %s' % __version__)
+    ap.add_argument('-v', '--verbose', nargs='?', metavar="NUM",
```
Why `metavar="NUM"` if the choices are level *labels*?

terencode replied:
Forgot to update after changing it.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
+    ap.add_argument('-v', '--verbose', nargs='?', metavar="NUM",
+                    choices=['DEBUG', 'INFO', 'WARNING', 'ERROR'], const='DEBUG', default='INFO',
+                    help="Set verbosity level. (choices: %(choices)s, default: %(default)s). "
+                    "Passing no arguments sets it to DEBUG ")
```
If this becomes a choice options, then the long form should probably be changed to `--verbosity` or `--log-level`.

SpotlightKid replied:
This is also an unrelated change and should go into a separate PR.

terencode replied:
I thought you'd be annoyed with too many PRs, that's why I combined everything in one...

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -69,6 +81,8 @@ def __init__(self, patterns, pattern_file=None, name=__program__, exact_matching
         self.connect_interval = connect_interval
         self.default_encoding = jacklib.ENCODING
 
+        jacklib.set_error_callback(self.error_callback)
```
This change, along the with changes to `jacklib.py` and the new `error_callback` method, should go into a separate PR.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
+            sys.exit("ERROR: Config file '{}' not found".format(path.realpath(args.config_file)))
+
 
+    logging.basicConfig(level=logging.getLevelName(args.verbose), format="%(levelname)s: %(message)s")
```
Since `args.verbose` is already guaranteed to be a valid log level name, and the `level` argument accepts a level name (since Python 3.2), the use of `logging.getLevelName` here makes no sense.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -347,6 +368,8 @@ def main(args=None):
     ap.add_argument('-p', '--pattern-file', metavar="FILE",
                     help="Read pattern pairs from FILE (one pattern per line)")
+    ap.add_argument('--config-file', metavar="FILE",
+                    help="Read program arguments from FILE")
```
A short option form (maybe `-C`?) would  nice.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -126,12 +140,15 @@ def add_patterns(self, ptn_output, ptn_input):
+                for ptn_output, ptn_input in pairwise(linefilter):
+                    self.add_patterns(ptn_output, ptn_input)
+        except FileNotFoundError:
+            sys.exit("ERROR: Pattern file '{}' not found".format(path.realpath(self.pattern_file)))
```
The exception this function can raise should be caught where it is called (i.e in the `__init__` method). If the function can do nothing to recover from the error, it should not catch it. Certainly it should not `sys.exit` from the middle of an object method.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
+                    else:
+                        log.error("Incorrect config option: '%s'", variable)
+        except FileNotFoundError:
+            sys.exit("ERROR: Config file '{}' not found".format(path.realpath(args.config_file)))
```
Config file parsing is an independent functionality and should go in to its own function.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -358,19 +381,37 @@ def main(args=None):
+                        if variable == "pattern_file" and not value.count('/'):
+                            # if pattern file path is not absolute, use the path of the config file
+                            value = "'{}/{}'".format(path.dirname(args.config_file), value.strip("\""))
+                        exec("args.%s = %s" % (variable, value))
```
Don't use `exec` (or `eval`)  to parse use supplied config files *ever*.

SpotlightKid commented on jackmatchmaker/__init__.py:
```diff
@@ -44,6 +45,17 @@ def pairwise(iterable):
+            vars.append(l[0].replace("-", "_"))
+            vars.append(l[1])
+
+    return pairwise(iter(vars))
```
This is unnecessarily complicated. Why not just append `name, value`  as a tuple in the first place?

SpotlightKid reviewed (commented):
Hi, I started this review and commented on a lot of details, but stopped at one point, because it became clear, that I'm not gonna accept this in the current form. I hope the existing comments are still helpful to you in some way.

SpotlightKid commented:
First off, thanks for your work an your PR.

However, I'm not gonna accept this.

* There are too many unrelated changes in one PR:
  * jack errors function
  * verbosity command line switch
  * exception handling in `add_patterns_from_file`
  * config file handling
* The main change, the config file handling is not implemented in a way that meets my quality standards.
* But, most importantly, I am not convinced, that adding config file handling to jack-matchmaker is necessary.
 
  * You can either just add the necessary options to the `ExecStart` line in the systemd file
  * Or you can add a shell wrapper, that just reads the command line options by, for example, sourcing another shell file.

  I'd like to explore the latter possibility, before I would consider adding config file parsing directly to jack-matchmaker.

Your work was not in vain, though, the jack error function stuff is valuable, if you'd like to make a separate PR for that, I'll be more receptive.  Same goes for the verbosity switch. You also uncovered that exceptions raised by `add_pattern_from_file` were not caught everywhere (which I just fixed in master).

terencode committed:
Add optional systemd unit along with environment file

Hence, no need for a config parser.

terencode commented:
Thanks for your detailed review.
I split the different change categories into different PR and applied your suggestions.
I gave the config parser a second thought and judged that it's indeed not needed If using the `EnvironmentFile` systemd variable.
I updated this PR to take this into account. It has a soft dependency on #9

SpotlightKid commented on jack-matchmaker.service:
```diff
@@ -0,0 +1,9 @@
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/usr/bin/jack-matchmaker -p $PATTERN_FILE -e $EXACT_MATCHING -N $CLIENT_NAME -m MAX_ATTEMPTS -v $VERBOSITY
```
What syntax is supported here? Does this support Bash variable expansion syntax e.g.

     ${PATTERN_FILE+-p $PATTERN_FILE}

so you only pass command line options that are defined in the environment file?

SpotlightKid replied:
Also, do these placeholders need quoting for values with whitespace in them?

terencode replied:
Parameter expansion works by passing the commands to `/bin/sh -c` and escaping the `$` with `$$` so systemd doesn't do its filtering.

Yes they do need quoting.

SpotlightKid commented on jack-matchmaker.service:
```diff
@@ -0,0 +1,9 @@
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/usr/bin/jack-matchmaker -p "$PATTERN_FILE" -e $EXACT_MATCHING -N "$CLIENT_NAME" -m $MAX_ATTEMPTS -v $VERBOSITY "$PATTERNS"
```
The `-e` option does not take arguments, so `-e $EXACT_MATCHING` will not work as intended.

The only way to set this I see is to have something like `EXTRA_OPTIONS="-e"` in the EnvironmentFile and then add `$EXTRA_OPTIONS` to the command line (not quoted!).

SpotlightKid replied:
`$PATTERNS` must not be quoted, so that the patterns are passed as separate command line arguments.

See section "Command Lines" [here](https://www.freedesktop.org/software/systemd/man/systemd.service.html).

terencode replied:
The `-e` option is fixed with parameter expansion: If EXACT_MATCHING is set to nothing it will not be presented as an argument.

SpotlightKid commented on jack-matchmaker.conf:
```diff
@@ -0,0 +1,11 @@
+PATTERN_FILE="/etc/jack-matchmaker/patterns.txt"
+PATTERNS=""
+CLIENT_NAME="jack-matchmaker"
+CONNECT_INTERVAL=3
```
This is missing in the command line (option `-I`).

terencode replied:
Added.

SpotlightKid commented on jack-matchmaker.service:
```diff
@@ -0,0 +1,9 @@
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/bin/sh -c "/usr/bin/jack-matchmaker $${PATTERN_FILE+-p \"$PATTERN_FILE\"} $${EXACT_MATCHING:+-e} $${CLIENT_NAME+-N \"$CLIENT_NAME\"} $${CONNECT_INTERVAL+-I $CONNECT_INTERVAL} $${MAX_ATTEMPTS+-m $MAX_ATTEMPTS} $${VERBOSITY+-v $VERBOSITY} $${PATTERNS+$PATTERNS}"
```
Uh, that's quite ugly, but it works :) Better to use `/bin/bash` instead of `/bin/sh`, though, since the extended variable expansion syntax is Bash-specific and some systems might not have `/bin/sh` linked to `/bin/bash`.

SpotlightKid replied:
If you use outer single quotes and just use `$$PATTERNS` (undefined variables just expand to an empty string), you can simplify the line slightly:

    ExecStart=/bin/bash -c '/usr/bin/jack-matchmaker $${PATTERN_FILE+-p "$PATTERN_FILE"} $${EXACT_MATCHING:+-e} $${CLIENT_NAME+-N "$CLIENT_NAME"} $${CONNECT_INTERVAL+-I $CONNECT_INTERVAL} $${MAX_ATTEMPTS+-m $MAX_ATTEMPTS} $${VERBOSITY+-v $VERBOSITY} $$PATTERNS'

terencode replied:
Thanks, updated :)

terencode replied:
~~Uh using single quotes doesn't seem to work.~~ Nevermind it works.

terencode replied:
Uh single quotes doesn't seem to work

terencode commented:
Here is an updated PKGBUILD that takes the new changes into account:
```
# Maintainer: Christopher Arndt <aur -at- chrisarndt -dot- de>

pkgname=jack-matchmaker-git
pkgver=0.7.1.r18.gcb49638
pkgrel=1
pkgdesc="A command line tool to auto-connect JACK ports matching given patterns. Git version."
arch=('any')
url="https://github.com/SpotlightKid/jack-matchmaker"
license=('GPL2')
depends=('python-setuptools' 'jack')
conflicts=('jack-matchmaker')
source=("jack-matchmaker-git::git+$url.git")
sha256sums=('SKIP')
backup=('etc/conf.d/jack-matchmaker')

pkgver() {
  cd "${srcdir}/${pkgname}"
  git describe --long --tags | sed 's/\([^-]*-g\)/r\1/;s/-/./g'
}


build() {
  cd "${srcdir}/${pkgname}"

  python setup.py build
}

package() {
  cd "${srcdir}/${pkgname}"

  python setup.py install --skip-build --root="${pkgdir}" --prefix=/usr \
    --optimize=1

  install -Dm 644 jack-matchmaker.conf $pkgdir/etc/conf.d/jack-matchmaker
  install -Dm 644 jack-matchmaker.service -t $pkgdir/usr/lib/systemd/user/
}
```
